### ISSUE TITLE
Memory usage spikes up when scraping from target exposing different series on every scrape (with series limits set)

### ISSUE THREAD

**Issue by blesswinsamuel** [VictoriaMetrics/VictoriaMetrics#3675]
### Describe the bug

This is a follow-up to the issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3660.

When I have vmagent scrape from an endpoint exposing different series on every scrape with series limits set and stream parse enabled (and with the series limit set to drop all new series), the memory usage spikes up during scrape and goes down once the scrape is done. The docs suggest that vmagent would use less memory with stream parse enabled, and I would expect the same.

In kubernetes environment with resource limits set, this memory spike often causes the pod to get OOMKilled.

### To Reproduce

To reproduce, use the similar setup described in the issue https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3660. In my case, I'm running it on a kubernetes cluster. I was able to reproduce the same locally on my Mac, but the profile generated showed zstd taking up a lot of memory and didn't show the other components clearly. So, I have uploaded the profiles from vmagent running on the Kubernetes cluster.

Whenever metrics are scraped from avalanche service, the memory goes up during scrape and then goes down once the scrape is done.

The default configuration of avalanche is set to emit entirely new series every 60s. So, with a scrape interval of 30s for avalanche, we will see memory spike every other scrape.

### Version

vmagent was built from the commit a844b9794247c2a9c6042dd1f6e1a0af8c4b19f2.

```
/ # ./vmagent-prod --version
vmagent-20230118-073318-heads-cluster-0-ga844b9794
```

### Profiles

- [Profile generated while scraping avalanche](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10449878/profile.10.pb.gz)
- [Profile generated while scraping avalanche - take 2](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10449882/profile.11.pb.gz)
- [Profile generated when not scraping avalanche](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10449887/profile.12.pb.gz)

Just want to mention that to get the first two profiles, I had to try many times to get it precisely at the point when the scrape was going on. I used the last scrape time displayed in the /targets endpoint to time it correctly. :)

### Screenshots

`sum(container_memory_working_set_bytes{container_name="vmagent"}) by (container_name) / 1e6` metric emitted by kube-state-metrics:

<img width="2892" alt="image" src="https://user-images.githubusercontent.com/815723/213273387-4ce3ef05-b3af-4624-b80a-24e383d2c79f.png">

Memory dump shows a large amount of memory is used by sendStaleSeries.

![image](https://user-images.githubusercontent.com/815723/213276249-021ed3b1-8925-4fea-8e15-839424835e40.png)

### Used command-line flags

Covered under the "To Reproduce" section.

### Additional information

After doing some tests, it looks like this memory spike happens when the target exposes a large number of new series it hasn't seen previously on every scrape.

**Comment by hagen1778** [comment]
Hello @blesswinsamuel! Could it be related to https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3668 ?
Would appreciate if you can verify it by using vmagent of version https://docs.victoriametrics.com/CHANGELOG.html#v1862

**Comment by blesswinsamuel** [comment]
Hi @hagen1778! I tried with the latest vmagent that was released yesterday. I see the same issue.

<img width="1839" alt="image" src="https://user-images.githubusercontent.com/815723/213458902-0e8e0d42-c149-4aa4-a56d-0f5db9fd8764.png">

[profile](https://github.com/VictoriaMetrics/VictoriaMetrics/files/10457259/profile.15.pb.gz)

It looks like the code before the [`if sw.seriesLimitExceeded` check](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/promscrape/scrapework.go#L805) in `sendStaleSeries` function is taking the most memory. Looks like disabling stale markers might help here. If nothing can be done to reduce memory usage with stale markers enabled in this scenario, please feel free to close this issue.

**Comment by hagen1778** [comment]
Hi! Thanks for the detailed issue.
Memory spike is indeed caused by sending the stale markers.
I think, there is a way to improve that so tagging the issue as an enhancement.

**Comment by valyala** [comment]
@blesswinsamuel , could you try building vmagent from the commit babecd83632446b6f24b57266def3782cf39c07d and verifying whether it helps removing memory usage spikes when the target changes the set of exposed metrics?

**Comment by blesswinsamuel** [comment]
@valyala @hagen1778, thanks for fixing this!

<img width="2888" alt="image" src="https://user-images.githubusercontent.com/815723/216020244-94b50138-2d2f-4a8a-a2d2-c920863b2ed3.png">

(scraping from avalanche emitting new set of 30k metrics on each scrape)

It seems to be much better now! Please feel free to close this issue.

**Comment by valyala** [comment]
@blesswinsamuel , thanks for the confirmation that the memory usage spikes are gone in vmagent now! The bugfix has been included in [vmagent v1.87.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). Closing the issue as fixed.
